### PR TITLE
Hide deprecation warnings inside derive expansions

### LINKED ISSUE #58822: Derives on deprecated items generate deprecation warnings
The following code generates a warning on nightly and beta but not on stable:
```rust
#[deprecated = "oh no"]
#[derive(Default)]
struct X;
```

### PR DESCRIPTION
**Pull request by oli-obk**
Fixes #58822

### CODE REVIEW

File: src/test/ui/deprecation/derive_on_deprecated_forbidden.rs [hunk 1]
```diff
@@ -0,0 +1,7 @@
+#![forbid(deprecated)]
+
+#[deprecated = "oh no"]
+#[derive(Default)]
+struct X; //~ ERROR deprecated
```
Comment (Centril) [inline src/test/ui/deprecation/derive_on_deprecated_forbidden.rs]: Note that erroring on `forbid` is a regression compared to stable. https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=c5255de60042c80607e6597f901295d8

Comment (oli-obk) [reply src/test/ui/deprecation/derive_on_deprecated_forbidden.rs]: I consider that a bugfix. If you are using forbid, you don't want any occurrances of a lint to go unnoticed.

Comment (oli-obk) [reply src/test/ui/deprecation/derive_on_deprecated_forbidden.rs]: OK, for the example here it's still just noise. Maybe I can just not emit deprecation warnings inside derive expansions

Comment (Centril) [reply src/test/ui/deprecation/derive_on_deprecated_forbidden.rs]: I suppose that's true :)

**Comment by petrochenkov** [comment]
I'd prefer to fix this in some other way.

`#[derive(Thing)]` should ideally be just `#[apply(Macro)]` with the macro performing all the expansion logic.
In this sense the `add_derived_markers` bit called from the generic derive infra is a wart that should be rather be refactored away than used more.

I think the usual approach "do not report warnings in macros from other crates" should work here, derive macros are always "from other crates" (there is a chance that built-in derives may need to be special-cased though).

**Comment by petrochenkov** [comment]
Alternatively, `impl Trait for Type { ... }` where `Type` is a deprecated type from the current crate is barely a *use*, but rather a part of `Type`s' *definition* even if it's not derived.
So, we may want to stop reporting such impls instead.

**Comment by Centril** [comment]
>  So, we may want to stop reporting such impls instead.

I don't think we should do this; deprecation may be part of an internal refactoring within a large application or crate. This happens e.g. in rustc with methods sometimes.

**Comment by oli-obk** [comment]
> I don't think we should do this; deprecation may be part of an internal refactoring within a large application or crate. This happens e.g. in rustc with methods sometimes.

The method would still get linted about if used from an impl for another type. It's not a crate wide "allow", but just an allow for impls.

**Comment by Centril** [comment]
I understand that the method would still be linted, but my point was not specific to methods, it applies equally to impl blocks, e.g.

```rust
#[deprecated]
struct Foo;

impl Foo { ... } // Not necessarily in the same file or part of a large file.
```

**Comment by oli-obk** [comment]
@petrochenkov would you be alright with just backporting the first commit, but putting the rest into nightly?

@Centril I removed the conroversial "don't lint on impls of deprecated types" change. The only functional changes now are

* fields of deprecated types in pattern matching don't get linted anymore
    * I believe this makes sense, since you already get a lint on mentioning the type during pattern matching
* some things like function calls are linted twice. Once for naming the function and once for calling the function. I can probably fix this if desired.

**Comment by Centril** [comment]
@oli-obk That's pretty good; the change to pattern matching seems like an UX improvement. Tho... for the purposes of beta-backport it seems a bit large. ^^

**Comment by petrochenkov** [comment]
@oli-obk 
>would you be alright with just backporting the first commit, but putting the rest into nightly?

Yeah, seems ok as a temporary measure for backport.

**Comment by petrochenkov** [comment]
@oli-obk 
Could you set the `S-waiting-on-review` label when this is ready?

**Comment by oli-obk** [comment]
Oh, right. Labels. Thanks for the reminder, I'll set those from now on.

**Comment by joshlf** [comment]
Sorry to be pushy, but is there an ETA on this? We have some work which is currently blocked on this being fixed (namely, we can't commit some code that exercises this bug). Happy to help if there's anything I can do to help move it along!

**Comment by petrochenkov** [comment]
I'll get to reviewing it this weekend or maybe tomorrow.

**Comment by joshlf** [comment]
Awesome thanks @petrochenkov ! Sorry again about being pushy.

**Comment by petrochenkov** [comment]
I should have reviewed this sooner, sorry.

To summarize, this PR does three things:
- Do not report deprecation in code from extern macros. This change fixes the original issue https://github.com/rust-lang/rust/issues/58822. It's also small and can be backported.
It can probably be submitted as a separate PR and quickly rolled up.
- Move deprecation checking from a separate stability pass to a lint pass.
I'd also want to see it moved, but to entirely different place, see the next comment (in progress...).
- Some other changes. It's difficult to see what is changed exactly when a lot of code is moved.

**Comment by petrochenkov** [comment]
Performing deprecation checking in the way it's performed now certainly has problems, multiple of which are shared with stability checking - import deprecations are ignored, intermediate path segments are not checked, macros are not checked for deprecation at all (https://github.com/rust-lang/rust/issues/49912).

The solution to all those problems is to organize stability/deprecation checking exactly like name privacy checking is organized now.
Basically, every time we resolve and entity to its `DefId` (this should ideally happens once for each entity) we immediately perform the checks associated with that `DefId`.
That means name privacy check are (and stability/deprecation checks should be) spread across the compilation process rather than performed in a single pass - for entities referring to macros or imports we check privacy/stability/deprecation during early name resolution, for entities using module-relative paths we check privacy/stability/deprecation during late name resolution, for entities resolved relatively to types (associated items including method calls, fields) we are checking privacy/stability/deprecation during type checking (see e.g. how calls to `check_stability` are spread through `rustc_typeck`).

The move of deprecation checking to a late lint pass doesn't make it closer to that goal, so if possible I'd prefer not to do that move and leave the code as is.

**Comment by oli-obk** [comment]
That makes a lot of sense. I implemented a small change that just skips the reporting in case the span points into a derive expansion.

**Comment by petrochenkov** [comment]
Thanks!
@bors r+

**Comment by pnkfelix** [comment]
discussed in T-compiler meeting. accepting for beta-backport.

**Pull request merged**